### PR TITLE
GH-36634: [Dev] Ensure merge script goes over all pages when requesting info from GitHub

### LINKED ISSUE #36634: [Dev] Merge PR script defaults to the wrong milestone when merging
### Describe the bug, including details regarding any error messages, version, and platform.

I saw that lots of PRs were merged with the wrong milestone (13.0.0) now that the maintenance branch has been created it should default to 14.0.0 but it is not as seen:
```
Would you like to update the associated issue? (y/n): y
Enter fix version [13.0.0]:
```
It does seem that the `12.0.0` and `12.0.1` maintenance branches are also not shown:
```
(Pdb) pr.maintenance_branches
['maint-0.11.x', 'maint-0.12.x', 'maint-0.14.x', 'maint-0.15.x', 'maint-0.17.x', 'maint-1.0.x', 'maint-3.0.x', 'maint-4.0.x', 'maint-6.0.x', 'maint-7.0.x', 'maint-7.0.1', 'maint-8.0.x', 'maint-9.0.0', 'maint-10.0.x', 'maint-10.0.0', 'maint-10.0.1', 'maint-11.0.0']
(Pdb) 
```
It might be due to pagination when requesting branches:
```
    def get_branches(self):
        return get_json("%s/branches" % (self.github_api),
                        headers=self.headers)
```

### Component(s)

Developer Tools

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

We currently were missing maintenance branches due to pagination on GH API.

### What changes are included in this PR?

Check whether the API is returning a paginated view and extend the list returned.

### Are these changes tested?

I have tested locally:

```
(Pdb) pr.maintenance_branches
['maint-0.11.x', 'maint-0.12.x', 'maint-0.14.x', 'maint-0.15.x', 'maint-0.17.x', 'maint-1.0.x', 'maint-3.0.x', 'maint-4.0.x', 'maint-6.0.x', 'maint-7.0.x', 'maint-7.0.1', 'maint-8.0.x', 'maint-9.0.0', 'maint-10.0.x', 'maint-10.0.0', 'maint-10.0.1', 'maint-11.0.0', 'maint-12.0.x', 'maint-12.0.0', 'maint-12.0.1', 'maint-13.0.0']
(Pdb) c
Enter fix version [14.0.0]: 
```
### Are there any user-facing changes?

No
* Closes: #36634

### CODE REVIEW

**Comment by raulcd** [comment]
There are currently no tests on the merge_pr script around neither GH issues or any interaction with GH. We should add them but this is tracked on this PR: https://github.com/apache/arrow/issues/14802

**Comment by raulcd** [comment]
Just for the record, other libraries like pygithub use the link header for the pagination and do a similar parsing: https://github.com/PyGithub/PyGithub/blob/de80ff4b914b9ffd47347d60b14cbf270e07f346/github/PaginatedList.py#L241

**Review by assignUser — APPROVED** [review]
Ah good catch! Is this something we need to add to archery somewhere too? Or not as that is using an api wrapper that handles it?

File: dev/merge_arrow_pr.py [hunk 1]
```diff
@@ -78,7 +78,26 @@ def get_json(url, headers=None):
...
+    next_responses = None
+    if "link" in response.headers:
+        links = response.headers['link'].split(', ')
+        next_url = None
+        for link in links:
+            if 'rel="next"' in link:
+                # Format: '<url>; rel="next"'
+                next_url = link.split(";")[0][1:-1]
+        if next_url:
+            next_responses = get_json(next_url, headers)
```
Comment (kou) [inline dev/merge_arrow_pr.py:93]: Can we simplify this?

```suggestion
        for link in links:
            if 'rel="next"' in link:
                # Format: '<url>; rel="next"'
                next_url = link.split(";")[0][1:-1]
                next_responses = get_json(next_url, headers)
```

File: dev/merge_arrow_pr.py [hunk 2]
```diff
@@ -78,7 +78,26 @@ def get_json(url, headers=None):
...
+    ret_val = response.json()
```
Comment (kou) [inline dev/merge_arrow_pr.py:94]: Can we use better name something like `responses`?

**Review by kou — APPROVED** [review]
+1

File: dev/merge_arrow_pr.py (hunk 2 above)
Comment (raulcd) [reply dev/merge_arrow_pr.py]: Yes, thanks for the review!

**Comment by raulcd** [comment]
> Is this something we need to add to archery somewhere too? Or not as that is using an api wrapper that handles it?

The pygithub and github3.py libraries used on archery handle pagination. I am not sure if there's any place where we should explicitly call to get next pages, etcetera but I don't think  so as  from what I've seen they just return iterators and handle that for us. Would be good though to try and refactor archery to use a single library so we are sure around those: https://github.com/apache/arrow/issues/33241

**Pull request merged**